docs/development_best_practices.md

```diff
@@ -0,0 +1,130 @@
+
+* For idempotent replay of data, the operator needs to store some meta-information for every window that would allow it to identify what data was sent in that window. This is called the idempotent state.
+	* If the external source of the input operator allows replayability, this could be information such as offset of last piece of data in the window, an identifier of the last piece of data itself or number of data tuples sent.
+	* However if the external source does not allow replayability from an operator specified point, then the entire data sent within the window may need to be persisted by the operator
+* The platform provides a utility called *WindowDataManager* to allow operators to save and retrieve idempotent state every window. Operators should use this to implement idempotency.
+
+## Output Operators
+
+Output operators typically connect to external storage systems such as filesystems, databases or key value stores to store data.
+
+* In some situations, the external systems may not be functioning in a reliable fashion. They may be having prolonged outages or performance problems. If the operator is being designed to work in such environments, it needs to be able to to handle these problems gracefully and not block the DAG or fail. In these scenarios the operator should cache the data into a local store such as HDFS and interact with external systems in a separate thread so as to not have problems in the operator lifecycle thread. The Reconciler pattern and operators that implement the pattern are available in the library for reference.
+
+### End-to-End Exactly Once
+
+When output operators store data in external systems, it is important that they do not lose data or write duplicate data when there is a failure event and the DAG recovers from that failure. In failure recovery, the windows from the previous checkpoint are replayed and the operator receives this data again. The operator should ensure that it does not write this data again. Operator developers should figure out how to do this specifically for the operators they are developing depending on the logic of the operators. Below are examples of how a couple of existing output operators do this for reference.
+
+* File output operator that writes data to files keeps track of the file lengths in the state. These lengths are checkpointed and restored on failure recovery. On restart, the operator truncates the file to the offset equal to the recovered length. This makes the data in the file same as it was at the time of checkpoint before the failure. The operator now writes the replayed data from the checkpoint in regular fashion as any other data. This ensures no data is lost or duplicated in the file.
```

Comment:
truncates the file to the offset equal to the recovered length => truncates the file to the recovered length